rmw_dds_common/src/qos.cpp

```diff
@@ -374,4 +384,262 @@ qos_profile_check_compatible(
+    publisher_profile->reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  }
+  if (RMW_QOS_POLICY_DURABILITY_BEST_AVAILABLE == publisher_profile->durability) {
+    publisher_profile->durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+  }
+
+  // Only use "manual by topic" liveliness if at least one  subscription is using manual by topic
+  // Use default deadline if all subscriptions have default deadline, otherwise use smallest
+  // Use default lease duration if all subscriptions have default lease, otherwise use smallest
+  bool use_manual_by_topic = false;
+  bool use_default_deadline = true;
+  rmw_time_t smallest_deadline = RMW_DURATION_INFINITE;
+  bool use_default_liveliness_lease_duration = true;
+  rmw_time_t smallest_liveliness_lease_duration = RMW_DURATION_INFINITE;
+  for (size_t i = 0u; i < subscriptions_info->size; ++i) {
+    const rmw_qos_profile_t & profile = subscriptions_info->info_array[i].qos_profile;
+    if (RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC == profile.liveliness) {
+      use_manual_by_topic = true;
+    }
+    if (profile.deadline != deadline_default) {
+      use_default_deadline = false;
+      if (profile.deadline < smallest_deadline) {
+        smallest_deadline = profile.deadline;
+      }
+    }
+    if (profile.liveliness_lease_duration != liveliness_lease_duration_default) {
+      use_default_liveliness_lease_duration = false;
+      if (profile.liveliness_lease_duration < smallest_liveliness_lease_duration) {
+        smallest_liveliness_lease_duration = profile.liveliness_lease_duration;
+      }
+    }
+  }
+
+  if (RMW_QOS_POLICY_LIVELINESS_BEST_AVAILABLE == publisher_profile->liveliness) {
+    if (use_manual_by_topic) {
+      publisher_profile->liveliness = RMW_QOS_POLICY_LIVELINESS_MANUAL_BY_TOPIC;
+    } else {
+      publisher_profile->liveliness = RMW_QOS_POLICY_LIVELINESS_AUTOMATIC;
+    }
+  }
+
+  if (deadline_best_available == publisher_profile->deadline) {
+    if (use_default_deadline) {
+      publisher_profile->deadline = RMW_QOS_DEADLINE_DEFAULT;
+    } else {
+      publisher_profile->deadline = smallest_deadline;
+    }
+  }
+
+  if (liveliness_lease_duration_best_available == publisher_profile->liveliness_lease_duration) {
+    if (use_default_liveliness_lease_duration) {
+      publisher_profile->liveliness_lease_duration = RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT;
+    } else {
+      publisher_profile->liveliness_lease_duration = smallest_liveliness_lease_duration;
+    }
+  }
+
+  return RMW_RET_OK;
+}
+
+static bool
+_qos_profile_has_best_available_policy(const rmw_qos_profile_t & qos_profile)
+{
+  if (RMW_QOS_POLICY_RELIABILITY_BEST_AVAILABLE == qos_profile.reliability) {
+    return true;
+  }
+  if (RMW_QOS_POLICY_DURABILITY_BEST_AVAILABLE == qos_profile.durability) {
+    return true;
+  }
+  if (RMW_QOS_POLICY_LIVELINESS_BEST_AVAILABLE == qos_profile.liveliness) {
+    return true;
+  }
+  if (deadline_best_available == qos_profile.deadline) {
```

Comment:
not that this is an error, but why using the conts here for `deadline_best_available` instead of just the `RMW_QOS_DEADLINE_BEST_AVAILABLE` like the comparison above?